sqlite.go

```diff
@@ -490,23 +490,22 @@ func (s *stmt) ExecContext(ctx context.Context, args []driver.NamedValue) (drive
 	if err := s.bindAll(args); err != nil {
 		return nil, s.reserr("Stmt.Exec(Bind)", err)
 	}
-	var done chan struct{}
+
+	// If non-nil, wait must be called prior to returning to ensure a
+	// cancellation context (if present) has completed, so that a cancellation
+	// cannot outlast this request and fire during a later execution.
+	var wait func()
 	if ctx.Value(queryCancelKey{}) != nil {
-		done = make(chan struct{})
+		done := make(chan struct{})
 		pctx, pcancel := context.WithCancel(ctx)
-		defer pcancel() // to make the AfterFunc fire and close(done)
-
 		db := s.stmt.DBHandle()
-		stop := context.AfterFunc(pctx, func() {
+		context.AfterFunc(pctx, func() {
 			defer close(done)
 			if ctx.Err() != nil {
 				db.Interrupt()
 			}
 		})
-		// In the event we get an error from the query's initial execution of
-		// sqlite3_step below and exit early, dissociate the cancellation since
-		// we don't want it to fire and potentially stop a later execution.
-		defer stop()
+		wait = func() { pcancel(); <-done }
```

Comment:
```
defer func() {
  pcancel()
  <-done
}()
```

would also work

Reply:
or if you prefer
```
wait := func() { pcancel(); <-done }
defer wait()
```

Reply:
fine as is too, just another approach

Reply:
It won't quite work the way you describe it—if we bind the cleanup function in this scope, we can't call it below, and it's essential we do so prior to returning (since, as you noted in Slack, the defer doesn't happen till too late)

Reply:
Per our offline conversation: Actually in this case a defer is fine, because the code to check the results doesn't depend on the cancellation being resolved. So, I swapped it to a defer (since that avoids having to think about scopes at all). Thanks!